crypto/evp_extra/p_kem.c

```diff
@@ -0,0 +1,233 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/evp.h>
+
+#include <openssl/err.h>
+#include <openssl/mem.h>
+
+#include "../fipsmodule/evp/internal.h"
+#include "../fipsmodule/delocate.h"
+#include "../kem/internal.h"
+#include "../internal.h"
+#include "internal.h"
+
+typedef struct {
+  const KEM *kem;
+} KEM_PKEY_CTX;
+
+static int pkey_kem_init(EVP_PKEY_CTX *ctx) {
+  KEM_PKEY_CTX *dctx;
+  dctx = OPENSSL_malloc(sizeof(KEM_PKEY_CTX));
+  if (dctx == NULL) {
+    return 0;
+  }
+  OPENSSL_memset(dctx, 0, sizeof(KEM_PKEY_CTX));
+
+  ctx->data = dctx;
+
+  return 1;
+}
+
+static int pkey_kem_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src) {
+  KEM_PKEY_CTX *dctx, *sctx;
+  if (!pkey_kem_init(dst)) {
+    return 0;
+  }
+  sctx = src->data;
+  dctx = dst->data;
+
+  dctx->kem = sctx->kem;
+
+  return 1;
+}
+
+static void pkey_kem_cleanup(EVP_PKEY_CTX *ctx) {
+  KEM_PKEY_CTX *dctx = ctx->data;
+  if (dctx == NULL) {
+    return;
+  }
+
+  OPENSSL_free(dctx);
+}
+
+static int pkey_kem_paramgen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
+  KEM_PKEY_CTX *dctx = ctx->data;
+  if (dctx->kem == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+    return 0;
+  }
+  KEM_KEY *key = KEM_KEY_new();
+  if (key == NULL ||
+      !KEM_KEY_init(key, dctx->kem)) {
+    KEM_KEY_free(key);
+    return 0;
+  }
+
+  EVP_PKEY_assign(pkey, EVP_PKEY_KEM, key);
+  return 1;
+}
+
+static int pkey_kem_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
+  KEM_PKEY_CTX *dctx = ctx->data;
+  const KEM *kem = dctx->kem;
+  if (kem == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    kem = KEM_KEY_get0_kem((KEM_KEY*)ctx->pkey->pkey.ptr);
+  }
+
+  KEM_KEY *key = KEM_KEY_new();
+  if (key == NULL ||
+      !KEM_KEY_init(key, kem) ||
+      !kem->method->keygen(key->public_key, key->secret_key) ||
+      !EVP_PKEY_assign(pkey, EVP_PKEY_KEM, key)) {
+    KEM_KEY_free(key);
+    return 0;
+  }
+
+  key->has_secret_key = 1;
+
+  return 1;
+}
+
+static int pkey_kem_encapsulate(EVP_PKEY_CTX *ctx,
+                                uint8_t *ciphertext,
+                                size_t  *ciphertext_len,
+                                uint8_t *shared_secret,
+                                size_t  *shared_secret_len) {
+  KEM_PKEY_CTX *dctx = ctx->data;
+  const KEM *kem = dctx->kem;
+  if (kem == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    kem = KEM_KEY_get0_kem((KEM_KEY*)ctx->pkey->pkey.ptr);
```

Comment:
Is 2 a valid state to end up in? Can you call pkey_kem_keygen if there is no specific kem configured?

Reply:
It is a valid state because the KEM can be configured in the PKEY object associated with the context. For example when creating a new raw key (see `RawKeyOperations` test in `evp_extra_tests.cc` for reference).
```
EVP_PKEY *pkey = EVP_PKEY_kem_new_raw_public_key(nid, pk, pk_len);
EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(pkey, NULL);

// you can use the context now for keygen/encaps/decaps.
```

BUT, to get back to your original question:
> Should this also assign it to ctx->data to avoid having to call get0 every time?

I just realized that it does actually assign it :)
```
const KEM *kem = ctx->data->kem;
  if (kem == NULL) {
    // ...
    kem = KEM_KEY_get0_kem(ctx->pkey->pkey.kem_key);
  }
```